File: Library/Classes/Core/Categories/UIKit/UICollectionView+CSExtension.m
//
// Created by Rene Dohan on 05/11/14.
// Copyright (c) 2014 Rene Dohan. All rights reserved.
//

#import "UICollectionView+CSExtension.h"
#import "NSObject+CSExtension.h"
#import "NSArray+CSExtension.h"
#import "UIView+CSExtension.h"
#import "UINib+CSExtension.h"
#import "UIView+CSDimension.h"
#import "UIView+CSLayout.h"

@implementation UICollectionView (CSExtension)

static NSString *const EMPTY_CELL = @"emptyCellIdentifier";
static NSString *const EMPTY_HEADER = @"emptyHeaderIdentifier";
static void *uiCollectionViewCellContent = &uiCollectionViewCellContent;

- (instancetype)construct {
    super.construct;
    self.backgroundColor = UIColor.clearColor;
    return self;
}

- (instancetype)setupCollection:(id <UICollectionViewDelegate, UICollectionViewDataSource>)parent {
    [self construct];
    self.delegate = parent;
    self.dataSource = parent;
    [self registerEmptyCell];
    [self reloadData];
    return self;
}

- (__kindof UICollectionViewCell *)dequeueCell:(NSString *)identifier :(NSIndexPath *)path {
    return [self dequeueReusableCellWithReuseIdentifier:identifier forIndexPath:path];
}

- (__kindof UICollectionViewCell *)dequeueCellForView:(Class)identifier :(NSIndexPath *)path {
    return [self dequeueReusableCellWithReuseIdentifier:[identifier description] forIndexPath:path];
}

- (instancetype)registerEmptyCell {
    [self registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:EMPTY_CELL];
    return self;
}

- (instancetype)registerEmptyCell:(Class)cellClass {
    [self registerClass:cellClass forCellWithReuseIdentifier:EMPTY_CELL];
    return self;
}

- (instancetype)registerEmptyHeader {
    [self registerClass:[UICollectionReusableView class] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:EMPTY_HEADER];
    return self;
}

- (UICollectionViewCell *)emptyCell:(Class)viewClass :(NSIndexPath *)path :(void (^)(UICollectionViewCell *))onCreate {
    var cell = [self dequeueCell:EMPTY_CELL :path];
    if (![cell.contentView.subviews.last getObject:uiCollectionViewCellContent]) {
        [cell.contentView add:[viewClass.construct setObject:uiCollectionViewCellContent :@(YES)]].matchParent;
        invokeWith(onCreate, cell);
    }
    return cell;
}

- (UICollectionReusableView *)dequeEmptyHeader:(NSIndexPath *)path {
    return [self dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader
                                    withReuseIdentifier:EMPTY_HEADER forIndexPath:path];
}


@end

File: Renetik/Renetik/Classes/Core/Extensions/UIKit/UICollectionView+CSExtension.swift
//
// Created by Rene on 2018-11-29.
//

import UIKit

private let DEFAULT_CELL_ID = "emptyCellIdentifier"

public extension UICollectionView {

    override open func construct() -> Self {
        super.construct()
        return background(.clear)
    }

    @discardableResult
    func construct(_ parent: UICollectionViewDelegate & UICollectionViewDataSource) -> Self {
        construct()
        delegates(parent)
        registerDefaultCell()
        reloadData()
        return self
    }

    @discardableResult
    func delegates(_ parent: UICollectionViewDelegate & UICollectionViewDataSource) -> Self {
        delegate = parent
        dataSource = parent
        return self
    }

    @discardableResult
    func register<CellType: UICollectionViewCell>(cell cellType: CellType.Type) -> Self {
        register(cellType, forCellWithReuseIdentifier: cellType.className())
        return self
    }

    func dequeue<CellType: UICollectionViewCell>(
            cell cellType: CellType.Type, _ path: IndexPath, onCreate: ((CellType) -> Void)? = nil) -> CellType {
        var cell = dequeueReusableCell(withReuseIdentifier: cellType.className(), for: path) as! CellType
        if cell.contentView.isEmpty {
//            cell.contentView.matchParent()
            cell.construct()
            onCreate?(cell)
        }
        return cell
    }

    public func scrollToPage() {
        var currentCellOffset = contentOffset
        currentCellOffset.x += width / 2
        var path = indexPathForItem(at: currentCellOffset)
        if path.isNil {
            currentCellOffset.x += 15
            path = indexPathForItem(at: currentCellOffset)
        }
        if path != nil {
            logInfo("Scrolling to page \(path!)")
            scrollToItem(at: path!, at: .centeredHorizontally, animated: true)
        }
    }

    func registerDefaultCell() -> Self {
        register(UICollectionViewCell.self, forCellWithReuseIdentifier: DEFAULT_CELL_ID)
        return self
    }

}
